app/models/chat_channel.rb

```diff
@@ -21,6 +22,30 @@ class ChatChannel < ActiveRecord::Base
             presence: true,
             allow_nil: true
 
+  scope :public_channels, -> do
+    where(chatable_type: public_channel_chatable_types)
+      .where("categories.id IS NOT NULL")
+      .joins(
+        "LEFT JOIN categories ON categories.id = chat_channels.chatable_id AND chat_channels.chatable_type = 'Category'",
+      )
+  end
+
+  class << self
+    def public_channel_chatable_types
+      ["Category"]
+    end
+
+    def chatable_types
+      public_channel_chatable_types << "DirectMessageChannel"
+    end
+  end
+
+  statuses.keys.each do |status|
```

Comment:
Used some simple meta-programming to define `open!` and the likes. This will also ensure those methods are up to date with the actual statuses.